src/comgr.cpp

```diff
@@ -0,0 +1,503 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2020 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include <miopen/env.hpp>
+#include <miopen/errors.hpp>
+#include <miopen/logger.hpp>
+#include <miopen/stringutils.hpp>
+#include <amd_comgr.h>
+#include <algorithm>
+#include <exception>
+#include <cstddef>
+#include <cstring>
+#include <tuple> // std::ignore
+#include <vector>
+
+/// \todo see issue #1222, PR #1316
+MIOPEN_DECLARE_ENV_VAR(MIOPEN_DEBUG_SRAM_EDC_DISABLED)
+
+#define DEBUG_DETAILED_LOG 0
+#define COMPILER_LC 1
+
+#if DEBUG_DETAILED_LOG
+#define EC_BASE_DETAILED_INFO(comgrcall, info) \
+    MIOPEN_LOG_I("Ok \'" #comgrcall "\' " << to_string(info))
+#else
+#define EC_BASE_DETAILED_INFO(comgrcall, info)
+#endif
+
+#define EC_BASE(comgrcall, info, expr2)                                   \
+    do                                                                    \
+    {                                                                     \
+        const amd_comgr_status_t status = (comgrcall);                    \
+        if(status != AMD_COMGR_STATUS_SUCCESS)                            \
+        {                                                                 \
+            MIOPEN_LOG_E("\'" #comgrcall "\' " << to_string(info) << ": " \
```

Comment:
Whats the point in logging if it is going to throw?

Reply:
This prints build log (that presumably contains error messages from compiler/linker etc) onto console.

MIOPEN_THROW does the same thing that we do here for offline compilation: https://github.com/ROCmSoftwarePlatform/MIOpen/blob/f1359cbc0a2cebcf7f6604bc1a2cdeddd6e8b0b3/src/hipoc/hipoc_program.cpp#L183-L184

MIOPEN_THROW should be moved to `HIPOCProgramImpl`. Thanks for noting this.

Reply:
Moved to proper place.

Reply:
Some explanatory comment added to the source code.